transcom/mymove
# B-18883: Allow Update Of ServiceItem That Is Linked To A Payment Request

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3a907739)

## Summary

This PR allows a service item to be updated if it is linked to a payment request ONLY if the following fields are updated: sitRequestedDelivery, sitCustomerContacted, or sitDepartureDate.

About the approach used:
Within my approach I created a method that is invoked within the checkPaymentRequest Method to check what allowable fields were changed and what unallowable fields were changed. The method returns true for itpasses the method/case or false. If it passes then the request will not be denied because of the case in which the serviceItem is linked to a payment request and only sitRequestedDelivery, sitCustomerContacted, or sitDepartureDate is being updated.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. If you do not have a move to use then create one as a servicemember
2. Work through the move approval flow if creating a new move
3. As a prime add service items to the move and approve them as the TOO
4. Submit a payment request as a prime. Take note of which service item you are adding to the payment request
5. As the prime attempt to update sitRequestedDelivery, sitCustomerContacted, or sitDepartureDate on the service item added in the payment request. In the prime sim you can only update these 3 anyways so it should pass. If it fails it may be because the shipment has no estimated weight in the DB.
6. Using postman or swagger update the serviceItem listed used in the payment request at http://primelocal:3000/prime/v1/mto-service-items/serviceItem-id. This time add a field other than sitRequestedDelivery, sitCustomerContacted, or sitDepartureDate. For example sitPostalCode. This request should be rejected.

### Backend
